snakemake/sourcecache.py

```diff
@@ -126,13 +126,24 @@ def __init__(
         self.repo_path = repo_path
         self.path = path
 
+    def get_basedir(self):
+        root_path, file_path, ref = split_git_path(self.get_path_or_uri())
+        file_name = "{}/{}".format(root_path, file_path).replace(root_path, "")
+        return LocalGitFile("file:/{}".format(root_path), file_name, ref=ref)
+
     def get_path_or_uri(self):
         return "git+{}/{}@{}".format(self.repo_path, self.path, self.ref)
 
     def join(self, path):
+        if isinstance(path, LocalSourceFile):
+            path = path.get_path_or_uri()
+        local_path = os.path.dirname(self.path).lstrip("/")
+        while path.startswith("../"):
+            path = path.lstrip("../")
+            local_path = os.path.dirname(local_path)
```

Comment:
Can you add some comments explaining why this is needed?

Reply:
we have a common.smk with the following format for schemas: 

https://github.com/snakemake-workflows/rna-seq-star-deseq2/blob/dfd0b9cb8c1ef8607505aef8baf885632227e8ac/workflow/rules/common.smk#L8-L10

which git.show didn't like since it generates a path looking like this 
`project/workflow/rules/../schemas/samples.schema.yaml"`

the code has been refactored and are now using os.path.normpath

which will change

`project/workflow/rules/../schemas/samples.schema.yaml"`

to 

`project/workflow/schemas/samples.schema.yaml"`